### ISSUE TITLE
Simplify css variables for theming purpose

### ISSUE THREAD

**Issue by jgoux** [material-components/material-components-web#754]
Hello,

I'm a React user and I'd like to use a CSS-in-JS solution to easily theme material-components-web.

Thanks to css variables, it shouldn't be too hard, so I started by extracting these from the dist css file (thanks `/--mdc-.+?: .+?;/g` :D) : 

```css
--mdc-theme-accent: #ff4081;
--mdc-theme-background: #fff;
--mdc-theme-primary: #3f51b5;
--mdc-theme-text-disabled-on-accent: rgba(255, 255, 255, 0.5);
--mdc-theme-text-disabled-on-background: rgba(0, 0, 0, 0.38);
--mdc-theme-text-disabled-on-dark: rgba(255, 255, 255, 0.5);
--mdc-theme-text-disabled-on-light: rgba(0, 0, 0, 0.38);
--mdc-theme-text-disabled-on-primary: rgba(255, 255, 255, 0.5);
--mdc-theme-text-hint-on-accent: rgba(255, 255, 255, 0.5);
--mdc-theme-text-hint-on-background: rgba(0, 0, 0, 0.38);
--mdc-theme-text-hint-on-dark: rgba(255, 255, 255, 0.5);
--mdc-theme-text-hint-on-light: rgba(0, 0, 0, 0.38);
--mdc-theme-text-hint-on-primary: rgba(255, 255, 255, 0.5);
--mdc-theme-text-icon-on-accent: rgba(255, 255, 255, 0.5);
--mdc-theme-text-icon-on-background: rgba(0, 0, 0, 0.38);
--mdc-theme-text-icon-on-dark: rgba(255, 255, 255, 0.5);
--mdc-theme-text-icon-on-light: rgba(0, 0, 0, 0.38);
--mdc-theme-text-icon-on-primary: rgba(255, 255, 255, 0.5);
--mdc-theme-text-primary-on-accent: white;
--mdc-theme-text-primary-on-background: rgba(0, 0, 0, 0.87);
--mdc-theme-text-primary-on-dark: white;
--mdc-theme-text-primary-on-light: rgba(0, 0, 0, 0.87);
--mdc-theme-text-primary-on-primary: white;
--mdc-theme-text-secondary-on-accent: rgba(255, 255, 255, 0.7);
--mdc-theme-text-secondary-on-background: rgba(0, 0, 0, 0.54);
--mdc-theme-text-secondary-on-dark: rgba(255, 255, 255, 0.7);
--mdc-theme-text-secondary-on-light: rgba(0, 0, 0, 0.54);
--mdc-theme-text-secondary-on-primary: rgba(255, 255, 255, 0.7);

/* Other variables not under the mdc-theme scope */
--mdc-dialog-dark-theme-bg-color: #303030;
--mdc-permanent-drawer-dark-theme-bg-color: #212121;
--mdc-persistent-drawer-dark-theme-bg-color: #212121;
--mdc-ripple-fg-scale: 1;
--mdc-ripple-fg-size: 0;
--mdc-ripple-fg-translate-end: 0;
--mdc-ripple-fg-translate-start: 0;
--mdc-ripple-left: 0;
--mdc-ripple-surface-height: 0;
--mdc-ripple-surface-width: 0;
--mdc-ripple-top: 0;
--mdc-toolbar-ratio-to-extend-flexible: 4;
```

Could you explain why all the variables aren't under the mdc-theme scope ?

Also, I'd like to apply [material design palettes](https://material.io/guidelines/style/color.html#color-color-tool) to select my primary and accent colors. What's the story between material-components-web and material design palettes ?

Wouldn't be interesting to put these as variables and use them in place of all the colors here ? 

Example taking the official red palette as primary color : 
```css
:root {
--mdc-theme-primary-50: #FFEBEE;
--mdc-theme-primary-100: #FFCDD2;
--mdc-theme-primary-200: #EF9A9A;
--mdc-theme-primary-300: #E57373;
--mdc-theme-primary-400: #EF5350;
--mdc-theme-primary-500: #F44336;
--mdc-theme-primary-600: #E53935;
--mdc-theme-primary-700: #D32F2F;
--mdc-theme-primary-800: #C62828;
--mdc-theme-primary-900: #B71C1C;
--mdc-theme-primary-A100: #FF8A80;
--mdc-theme-primary-A200: #FF5252;
--mdc-theme-primary-A400: #FF1744;
--mdc-theme-primary-A700: #D50000;
}
```

Then you could use that to color all the primary components.
I started to experiment the palette generation using only the `500` color : https://codesandbox.io/s/kZN3kNpQK

Also, I'm pretty sure all the `mdc-theme-text-*` variables could be derived from the primary/secondary color (using constrast calculation). Do you intend to provide a javascript helper or any documentation about this ?

To sum up, I'd like to be able to theme a whole application providing only one primary color and one accent color. I think all the other colors could be derived from that.

**Comment by Garbee** [comment]
> Could you explain why all the variables aren't under the mdc-theme scope ?

Theme vars are set under the `theme` package. While the others correspond to their package names. `mdc` is the prefix followed by the package name, then the actual var name. Having `mdc-theme` before every given variable seems unnecessary.

> Wouldn't be interesting to put these as variables (color palette) and use them in place of all the colors here ?

iirc we didn't do this for MCW since there was some discussion of the color scheme possibly changing dramatically in the future. So we went with the most minimal configuration we could to be capable of adapting more easily in case of spec changes in the future.

> Also, I'm pretty sure all the `mdc-theme-text-*` variables

They are derived in the [Sass build](https://github.com/material-components/material-components-web/blob/master/packages/mdc-theme/_functions.scss). This could possibly be ported to a JS implementation. However, I'm unsure of whether that should reside within MCW since it isn't something required by us in any way. It seems like an application level detail which means that is best done by the app developer or as a community package that can be required in as-needed.

**Comment by traviskaufman** [comment]
Going off of what @Garbee said, we do have work queued up to create a sass/css helper library for our colors (#27) , as well as a chore to allow programmatic theming of our components via JS (#3), however we haven't yet gotten around to these (you can check out our [public pivotal tracker](https://www.pivotaltracker.com/n/projects/1664011), which mirrors our GH issues, to see what we're currently working on). 

In the meantime, you can theme our application using only primary/secondary colors using Sass. For React/CSS-in-JS, you could do this by using webpack's sass-loader and then try embedding it via something like https://www.npmjs.com/package/inline-css-webpack-loader. Our [theming guide]() has more info on how to theme using Sass: https://github.com/material-components/material-components-web/blob/master/docs/theming.md#step-3-changing-the-theme-with-sass
